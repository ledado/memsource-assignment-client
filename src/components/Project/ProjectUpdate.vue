<template>
  <v-row no-gutters justify="center">
    <v-col cols="4">
      <v-row>
        <v-col cols="12" class="text-h4">
          {{ $t('project.form.update.title') }}
        </v-col>
        <v-col cols="12" v-if="projectDetailLoading" class="text-center my-4">
          <v-progress-circular indeterminate color="primary"></v-progress-circular>
        </v-col>
        <v-col cols="12" v-if="projectToUpdate">
          <project-form
            ref="projectForm"
            v-model="projectToUpdate"
            :type="ProjectFormTypeEnum.UPDATE"
            :loading="loading"
            @submit="updateProject"
          ></project-form>
        </v-col>
        <v-col cols="12" v-if="!projectToUpdate && !projectDetailLoading" class="text-center text-h5">
          {{ $t('project.detail.not_found') }}
        </v-col>
        <v-col cols="12">
          <v-btn color="error" :to="{name: $data.routerTypes.MAIN}" block>
            {{ $t('global.cancel') }}
          </v-btn>
        </v-col>
        <v-col cols="12">
          <v-btn color="primary" :to="{name: $data.routerTypes.MAIN}" block>
            {{ $t('project.form.back') }}
          </v-btn>
        </v-col>
      </v-row>
    </v-col>
  </v-row>
</template>

<script lang="ts">
import {Component, Prop, Vue} from 'vue-property-decorator';
import ProjectService from '@/service/api/ProjectService';
import {ProjectItem, ProjectFormTypeEnum} from '@/service/api/types/ProjectTypes';
import ProjectForm from '@/components/Project/ProjectForm.vue';

@Component({
  components: {ProjectForm},
})
export default class ProjectUpdate extends Vue {
  @Prop() projectId!: string;
  ProjectFormTypeEnum = ProjectFormTypeEnum;

  loading = false;
  projectDetailLoading = false;
  projectToUpdate: ProjectItem | null = null;

  $refs!: {
    projectForm: any;
  };

  created() {
    this.projectDetailLoading = true;
    ProjectService.getDetail(this.projectId)
      .then((response) => {
        this.projectToUpdate = response.data;
      })
      .catch((error) => {
        if (error.response.status !== 404) {
          this.$toasted.error(this.$t('global.error.request').toString());
        }
      })
      .finally(() => {
        this.projectDetailLoading = false;
      });
  }

  updateProject() {
    this.loading = true;
    const date = new Date();
    ProjectService.updateProject({
      ...this.projectToUpdate!,
      dateUpdated: date.toISOString(),
    })
      .then(() => {
        this.$toasted.success(this.$t('project.form.update.success').toString());
      })
      .catch(() => {
        this.$toasted.error(this.$t('global.error.request').toString());
      })
      .finally(() => {
        this.loading = false;
      });
  }
}
</script>
<style lang="scss" scoped></style>
