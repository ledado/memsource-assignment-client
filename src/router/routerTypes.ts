export const routerTypes = {
  MAIN: 'main',
  PROJECT_NEW: 'project-new',
  PROJECT_UPDATE: 'project-update',
};
