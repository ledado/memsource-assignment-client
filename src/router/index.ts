import Vue from 'vue';
import VueRouter, {RouteConfig} from 'vue-router';
import {errorBus} from '@/service/events/ErrorBus';
import {layoutTypes} from '@/layouts/layoutTypes';
import {routerTypes} from '@/router/routerTypes';

Vue.use(VueRouter);

const routes: Array<RouteConfig> = [
  {
    path: '/',
    name: routerTypes.MAIN,
    component: () => import('../views/Main.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
  },
  {
    path: '/project/new',
    name: routerTypes.PROJECT_NEW,
    component: () => import('../views/Project/MainProjectCreate.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
  },
  {
    path: '/project/:id',
    name: routerTypes.PROJECT_UPDATE,
    component: () => import('../views/Project/MainProjectUpdate.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
    props: true,
  },
  {
    path: '*',
    component: () => import('../views/Error/Error404.vue'),
  },
];

const router = new VueRouter({
  mode: 'history',
  base: process.env.BASE_URL,
  routes,
});

// error state clean up
router.afterEach((to, from) => {
  if (from.name && errorBus.state) {
    errorBus.errorCode = undefined;
    errorBus.state = false;
  }
});

export default router;
