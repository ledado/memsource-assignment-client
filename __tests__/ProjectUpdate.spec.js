import {shallowMount} from '@vue/test-utils';
import ProjectUpdate from '@/components/Project/ProjectUpdate.vue';
import {routerTypes} from '@/router/routerTypes';

describe('ProjectUpdate.vue', () => {
  it('render component', () => {
    const wrapper = shallowMount(ProjectUpdate, {
      propsData: {
        projectId: 10,
      },
      stubs: ['router-link'],
      mocks: {
        $t: (msg) => msg,
      },
      mixins: [
        {
          data: () => {
            return {
              routerTypes,
            };
          },
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
