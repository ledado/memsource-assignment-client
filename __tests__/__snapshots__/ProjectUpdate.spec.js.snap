// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectUpdate.vue render component 1`] = `
"<v-row-stub tag=\\"div\\" nogutters=\\"true\\" justify=\\"center\\">
  <v-col-stub cols=\\"4\\" tag=\\"div\\">
    <v-row-stub tag=\\"div\\">
      <v-col-stub cols=\\"12\\" tag=\\"div\\" class=\\"text-h4\\">
        project.form.update.title
      </v-col-stub>
      <v-col-stub cols=\\"12\\" tag=\\"div\\" class=\\"text-center my-4\\">
        <v-progress-circular-stub color=\\"primary\\" indeterminate=\\"true\\" rotate=\\"0\\" size=\\"32\\" width=\\"4\\" value=\\"0\\"></v-progress-circular-stub>
      </v-col-stub>
      <!---->
      <!---->
      <v-col-stub cols=\\"12\\" tag=\\"div\\">
        <v-btn-stub color=\\"error\\" tag=\\"button\\" activeclass=\\"\\" to=\\"[object Object]\\" block=\\"true\\" type=\\"button\\">
          global.cancel
        </v-btn-stub>
      </v-col-stub>
      <v-col-stub cols=\\"12\\" tag=\\"div\\">
        <v-btn-stub color=\\"primary\\" tag=\\"button\\" activeclass=\\"\\" to=\\"[object Object]\\" block=\\"true\\" type=\\"button\\">
          project.form.back
        </v-btn-stub>
      </v-col-stub>
    </v-row-stub>
  </v-col-stub>
</v-row-stub>"
`;
